editoast/src/models/stdcm_search_environment.rs

```diff
@@ -30,46 +28,54 @@ pub struct StdcmSearchEnvironment {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub work_schedule_group_id: Option<i64>,
     pub timetable_id: i64,
+    /// The start of the search time window.
+    /// Usually, trains schedules from the `timetable_id` runs within this window.
     pub search_window_begin: DateTime<Utc>,
+    /// The end of the search time window.
     pub search_window_end: DateTime<Utc>,
     #[schema(nullable = false)]
     #[serde(skip_serializing_if = "Option::is_none")]
     pub temporary_speed_limit_group_id: Option<i64>,
+    /// The time window start point where the environment is enabled.
+    pub enabled_from: DateTime<Utc>,
+    /// The time window end point where the environment is enabled.
+    /// This value is usually lower than the `search_window_begin`, since a search is performed before the train rolls.
+    pub enabled_to: DateTime<Utc>,
 }
 
 impl StdcmSearchEnvironment {
-    /// Retrieve the latest search environment. Returns None if no search environment is found.
-    pub async fn retrieve_latest(conn: &mut DbConnection) -> Option<Self> {
+    /// Retrieve the enabled search environment. If no env is enabled returns the most recent `enabled_to`.
+    /// In case of multiple enabled environments, the one with the highest `id` is returned.
+    pub async fn retrieve_latest_enabled(conn: &mut DbConnection) -> Option<Self> {
         use editoast_models::tables::stdcm_search_environment::dsl::*;
+        // Search for enabled env
+        let enabled_env = stdcm_search_environment
+            .order_by(id.desc())
+            .filter(enabled_from.le(diesel::dsl::now))
+            .filter(enabled_to.ge(diesel::dsl::now))
+            .first::<Row<StdcmSearchEnvironment>>(conn.write().await.deref_mut())
+            .await
+            .map(Into::into)
+            .ok();
+        if enabled_env.is_some() {
+            return enabled_env;
+        }
+
+        // Search for the most recent env
+        tracing::warn!("No STDCM search environment enabled");
         stdcm_search_environment
-            .order_by((search_window_end.desc(), search_window_begin.asc()))
+            .order_by((enabled_to.desc(), id.desc()))
             .first::<Row<StdcmSearchEnvironment>>(conn.write().await.deref_mut())
             .await
             .map(Into::into)
             .ok()
     }
```

Comment:
nitpick: Maybe this would be more idiomatic?

```rust
enabled_env.or_else(|| {
    tracing::warn!("No STDCM search environment enabled");
    stdcm_search_environment
        .order_by((enabled_to.desc(), id.desc()))
        .first::<Row<StdcmSearchEnvironment>>(conn.write().await.deref_mut())
        .await
        .map(Into::into)
        .ok()
})

Reply:
It requires an async closure. So IMO it's simpler as it is.